pallets/parachain-staking/src/lib.rs

```diff
@@ -2116,40 +2238,65 @@ pub mod pallet {
 	{
 		/// Compute coinbase rewards for block production and distribute it to
 		/// collator's (block producer) and its delegators according to their
-		/// stake.
+		/// stake and the current InflationInfo.
+		///
+		/// The rewards are split between collators and delegators with
+		/// different reward rates and maximum staking rates. The latter is
+		/// required to have at most our targeted inflation because rewards are
+		/// minted. Rewards are immediately available without any restrictions
+		/// after minting.
+		///
+		/// If the current staking rate is below the maximum, each collator and
+		/// delegator receives the corresponding `reward_rate * stake /
+		/// blocks_per_year`. Since a collator can only author blocks every
+		/// `MaxSelectedCandidates` many rounds, we multiply the reward with
+		/// this number. As a result, a collator who has been in the set of
+		/// selected candidates, eventually receives `reward_rate * stake` after
+		/// one year.
+		///
+		/// However, if the current staking rate exceeds the max staking rate,
+		/// the reward will be reduced by `max_rate / current_rate`. E.g., if
+		/// the current rate is at 50% and the max rate at 40%, the reward is
+		/// reduced by 20% by multiplying the reward with 80%.
 		///
 		/// # <weight>
 		/// Weight: O(D) where D is the number of delegators of this collator
 		/// block author bounded by `MaxDelegatorsPerCollator`.
-		/// - Reads: AtStake, Total, Balance
+		/// - Reads: CollatorState, Total, Balance, InflationConfig,
+		///   MaxSelectedCandidates
 		/// - Writes: D * Balance
 		/// # </weight>
 		fn note_author(author: T::AccountId) {
-			let state = <AtStake<T>>::get(author.clone());
-			if state.stake >= T::MinCollatorStk::get() && state.total >= T::MinCollatorStk::get() {
-				let TotalStake {
-					collators: total_collators,
-					delegators: total_delegators,
-				} = <Total<T>>::get();
-				let (c_rewards, d_rewards) = Self::compute_block_issuance(total_collators, total_delegators);
-
-				let amt_due_collator = c_rewards;
-				let delegator_stake = state.total.saturating_sub(state.stake);
-				let amt_due_delegators = d_rewards;
-
-				// Reward collator
-				Self::do_reward(&author, amt_due_collator);
-
-				// Reward delegators
-				// Reward delegators due portion
-				for Stake { owner, amount } in state.delegators {
-					if amount >= T::MinDelegatorStk::get() {
-						// Compare this delegator's stake with the total amount of
-						// delegated stake for this collator
-						// multiplication with perquintill cannot overflow
-						let percent = Perquintill::from_rational(amount, delegator_stake);
-						let due = percent * amt_due_delegators;
-						Self::do_reward(&owner, due);
+			// should always include state
+			if let Some(state) = <CollatorState<T>>::get(author.clone()) {
+				if state.stake >= T::MinCollatorStk::get() && state.total >= T::MinCollatorStk::get() {
```

Comment:
do we need to check this here? This should be ensured in the extrinsics? If these preconditions are not met, what is the outcome here? At the moment the block will be produced without rewards. if we remove the check, the rewards are just less? Or can the function go to infinity if the total stake is to low?

Reply:
Yeah, the check can be removed.